### PR TITLE
feat(http): URLSearchParams accepts object literal

### LINKED ISSUE #7370: Poposal: URLSearchParams - additional constructor with object param
Hello,
Why there is no constructor which allows us to pass Object as parameter in URLSearchParams class ?
At least there is nothing documented in the api yet.

I can navigate by routes like:
`router.parent.navigate(["Users", {"id": "10", "name" : "John,Martin"});`

Then if I inject RouteParams to component constructor I can read all params like:

```
params = routeParams.params
// params =  {"id": "10", "name" : "John,Martin"}
```

But I can't simply pass my routeParams to http request and I have to iterate through all params and add them manually using set method like in example below:

``` TS
    getData(params) { // params =  {"id": "10", "name" : "John,Martin"}
        let searchParams = new URLSearchParams();
        for (let param in params) {
            searchParams.set(param, params[param]);
        }

        let options = new RequestOptions({
            search: searchParams
        });

        return this.http.get(this.myUrl, options)
            .map(res => <MyClass[]> res.json())
            .catch(this.handleError)
    }
```

It is nothing, but in my opinion it would be great if it were available additional constructor of URLSearchParams class:
`constructor(objParams?: Object)`
or additional string property in RouteParams class which will store route search query like:
`"?id=10&name=John,Martin"`

Am I right ? Maybe I'm doing something wrong and there is a better way to pass search query from url to http request ?

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Fixes https://github.com/angular/angular/issues/7370 / https://github.com/angular/angular/issues/8858

### CODE REVIEW

File: modules/@angular/http/src/url_search_params.ts [hunk 1]
```diff
@@ -81,8 +96,13 @@ function standardEncoding(v: string): string {
 export class URLSearchParams {
   paramsMap: Map<string, string[]>;
   constructor(
-      public rawParams: string = '', private queryEncoder: QueryEncoder = new QueryEncoder()) {
-    this.paramsMap = paramParser(rawParams);
+      public rawParams: string|{[key: string]: any} = '',
+      private queryEncoder: QueryEncoder = new QueryEncoder()) {
+    if (typeof rawParams === 'string') {
+      this.paramsMap = strParamParser(rawParams);
```
Comment (vicb) [inline modules/@angular/http/src/url_search_params.ts]: return to `parameterParser` and move the check there please.

File: modules/@angular/http/src/url_search_params.ts [hunk 2]
```diff
@@ -21,6 +21,21 @@ function paramParser(rawParams: string = ''): Map<string, string[]> {
   }
   return map;
 }
+
+function objParamParser(rawParams: {[key: string]: any} = {}): Map<string, string[]> {
+  const map = new Map<string, string[]>();
+  Object.keys(rawParams).forEach((key: string) => {
+    let list: string[] = map.get(key) || [];
```
Comment (vicb) [inline modules/@angular/http/src/url_search_params.ts]: `const` how could we have a `map.get(key)` ?

File: modules/@angular/http/src/url_search_params.ts [hunk 3]
```diff
@@ -6,21 +6,35 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-function paramParser(rawParams: string = ''): Map<string, string[]> {
+function paramParser(rawParams: string | {[key: string]: any}): Map<string, string[]> {
   const map = new Map<string, string[]>();
```
Comment (vicb) [inline modules/@angular/http/src/url_search_params.ts]: rename map to params

File: modules/@angular/http/src/url_search_params.ts [hunk 4]
```diff
@@ -6,21 +6,35 @@
...
-  if (rawParams.length > 0) {
-    const params: string[] = rawParams.split('&');
-    params.forEach((param: string) => {
-      const eqIdx = param.indexOf('=');
-      const [key, val]: string[] =
-          eqIdx == -1 ? [param, ''] : [param.slice(0, eqIdx), param.slice(eqIdx + 1)];
-      const list = map.get(key) || [];
-      list.push(val);
+  if (typeof rawParams === 'string') {
+    if (rawParams.length > 0) {
```
Comment (vicb) [inline modules/@angular/http/src/url_search_params.ts]: what if length = 0 ?

File: modules/@angular/http/src/url_search_params.ts [hunk 5]
```diff
@@ -6,21 +6,35 @@
...
+      params.forEach((param: string) => {
+        const eqIdx = param.indexOf('=');
+        const [key, val]: string[] =
+            eqIdx == -1 ? [param, ''] : [param.slice(0, eqIdx), param.slice(eqIdx + 1)];
+        const list = map.get(key) || [];
+        list.push(val);
+        map.set(key, list);
+      });
+    }
+  } else {
```
Comment (vicb) [inline modules/@angular/http/src/url_search_params.ts]: return here and remove the else

File: modules/@angular/http/src/url_search_params.ts [hunk 6]
```diff
@@ -6,21 +6,35 @@
...
+    Object.keys(rawParams).forEach((key: string) => {
+      let list: string[];
+      let value: any = rawParams[key];
+      if (Array.isArray(value)) {
+        list = value.map((item: any) => JSON.stringify(item));
```
Comment (vicb) [inline modules/@angular/http/src/url_search_params.ts]: what if string items ?

File: modules/@angular/http/test/url_search_params_spec.ts [hunk 7]
```diff
@@ -167,5 +167,13 @@ export function main() {
       expect(params.toString()).toEqual('q=Q');
     });
 
+    it('constructor should accept object literal', () => {
```
Comment (vicb) [inline modules/@angular/http/test/url_search_params_spec.ts]: literals

File: modules/@angular/http/src/url_search_params.ts (hunk 4 above)
Comment (DzmitryShylovich) [inline modules/@angular/http/src/url_search_params.ts]: it will return an empty map. there're test for this case

**Comment by DzmitryShylovich** [comment]
can u restart ci pls? failed because of timeout

File: modules/@angular/http/src/url_search_params.ts [hunk 8]
```diff
@@ -6,21 +6,51 @@
...
   if (rawParams.length > 0) {
-    const params: string[] = rawParams.split('&');
-    params.forEach((param: string) => {
+    const list: string[] = rawParams.split('&');
+    list.forEach((param: string) => {
       const eqIdx = param.indexOf('=');
       const [key, val]: string[] =
           eqIdx == -1 ? [param, ''] : [param.slice(0, eqIdx), param.slice(eqIdx + 1)];
-      const list = map.get(key) || [];
+      const list = params.get(key) || [];
```
Comment (vicb) [inline modules/@angular/http/src/url_search_params.ts]: nit: list -> value (there is already a list outside of the loop)

Comment (vicb) [inline modules/@angular/http/src/url_search_params.ts]: I would be fine with `p.set(name, (p.get(name) || []).concat(value))`

File: modules/@angular/http/src/url_search_params.ts [hunk 9]
```diff
@@ -6,21 +6,51 @@
...
+
+function parseStrParams(rawParams: string): Map<string, string[]> {
+  const params = new Map<string, string[]>();
   if (rawParams.length > 0) {
-    const params: string[] = rawParams.split('&');
-    params.forEach((param: string) => {
+    const list: string[] = rawParams.split('&');
+    list.forEach((param: string) => {
       const eqIdx = param.indexOf('=');
       const [key, val]: string[] =
```
Comment (vicb) [inline modules/@angular/http/src/url_search_params.ts]: [name, val]

**Review by vicb — CHANGES_REQUESTED** [review]

File: modules/@angular/http/src/url_search_params.ts [hunk 10]
```diff
@@ -6,21 +6,51 @@
...
+    let list: string[];
+    const value: any = rawParams[key];
+    if (Array.isArray(value)) {
+      list = value.map((item: any) => {
+        if (typeof item !== 'string') {
+          item = JSON.stringify(item);
+        }
+        return item;
+      });
+    } else if (typeof value !== 'string') {
```
Comment (vicb) [inline modules/@angular/http/src/url_search_params.ts]: could you make === string first ?

**Comment by laco0416** [comment]
Really should this be supported as URLSearchParams constructor? `URLSearchParams` is not a feature introduced only for Angular, but it's a shim for the [standard spec](https://url.spec.whatwg.org/#urlsearchparams) , I think. 
Current constructor `URLSearchParams(optional string, optional QueryEncoder)` is still compatible with that spec. If this PR has accepted, `@angular/http` will lose the documented stance below; 

> ... based on [URLSearchParams](https://url.spec.whatwg.org/#urlsearchparams) in the url living standard,

**Comment by robwormald** [comment]
Agree with @laco0416 - we should not change the behavior of URLSearchParams, as ideally we'd drop our own implementation and rely on the native.

**Comment by vicb** [comment]
Makes sense, closing then

**Comment by vkrol** [comment]
The specification was updated recently https://github.com/whatwg/url/commit/31ddc5bd0c4ca1b6d837029c926ec87bc472c168

**Pull request closed without merging**